src/var.c

```diff
@@ -0,0 +1,247 @@
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ */
+
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#include <ctype.h>
+
+#include <rdma/fi_errno.h>
+#include <rdma/fi_var.h>
+#include <rdma/fi_log.h>
+
+struct fi_var_t {
+	const struct fi_provider *provider;
+	char *var_name;
+	char *help_string;
+
+	char *env_var_name;
+
+	struct fi_var_t *next;
+};
+
+static struct fi_var_t *fi_vars = NULL;
+
+
+int fi_var_register(const struct fi_provider *provider, const char *var_name,
+		const char *help_string)
+{
+	int i;
+	struct fi_var_t *v;
+
+	// Check for bozo cases
+	if (provider == NULL || var_name == NULL || help_string == NULL ||
+		*help_string == '\0') {
+		FI_DBG(provider, FI_LOG_CORE,
+			"Failed to register %s variable: provider coding error\n",
+			var_name);
+		return -FI_EINVAL;
+	}
+
+	v = calloc(1, sizeof(*v));
+	if (!v) {
+		FI_DBG(provider, FI_LOG_CORE,
+			"Failed to register %s variable: ENOMEM\n", var_name);
+		return -FI_ENOMEM;
+	}
+
+	v->provider = provider;
+	v->var_name = strdup(var_name);
+	v->help_string = strdup(help_string);
+	asprintf(&v->env_var_name, "FI_%s_%s",
+		v->provider->name,
+		v->var_name);
+	if (!v->var_name || !v->help_string || !v->env_var_name) {
+		free(v);
+		FI_DBG(provider, FI_LOG_CORE,
+			"Failed to register %s variable: ENOMEM\n", var_name);
+		return -FI_ENOMEM;
+	}
+
+	for (i = 0; v->env_var_name[i]; ++i)
+		v->env_var_name[i] = toupper(v->env_var_name[i]);
+
+	v->next = fi_vars;
+	fi_vars = v;
+
+	FI_INFO(provider, FI_LOG_CORE, "registered var %s\n", var_name);
+
+	return FI_SUCCESS;
+}
+
+static int fi_var_get(struct fi_provider *provider, const char *var_name,
+		char **value)
+{
+	struct fi_var_t *v;
+
+	// Check for bozo cases
+	if (var_name == NULL || value == NULL) {
+		FI_DBG(provider, FI_LOG_CORE,
+			"Failed to read %s variable: provider coding error\n",
+			var_name);
+		return -FI_EINVAL;
+	}
+
+	for (v = fi_vars; v; v = v->next) {
+		if (strcmp(v->provider->name, provider->name) == 0 &&
+			strcmp(v->var_name, var_name) == 0) {
+			*value = getenv(v->env_var_name);
+
+			return FI_SUCCESS;
+		}
+	}
+
+	FI_DBG(provider, FI_LOG_CORE,
+		"Failed to read %s variable: was not registered\n", var_name);
+	return -FI_ENOENT;
+}
+
+int fi_var_get_str(struct fi_provider *provider, const char *var_name,
+		char **value)
+{
+	int ret;
+
+	ret = fi_var_get(provider, var_name, value);
+	if (ret == FI_SUCCESS) {
+		if (*value) {
+			FI_INFO(provider, FI_LOG_CORE,
+				"read string var %s=%s\n", var_name, *value);
+			return FI_SUCCESS;
+		} else {
+			FI_INFO(provider, FI_LOG_CORE,
+				"read string var %s=<not set>\n", var_name);
+			return -FI_ENODATA;
+		}
```

Comment:
You can drop the 'return FI_SUCCESS' line and just let the code fall through.  And change 'return -FI_ENODATA' to 'ret = -FI_ENODATA', which will give us a single exit point.  This same flow can be followed below.